plugins/linux-capture/pipewire.c

```diff
@@ -0,0 +1,1110 @@
+		call->request_path, "org.freedesktop.portal.Request", "Close",
+		NULL, NULL, G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL, NULL);
+}
+
+static dbus_call_data *subscribe_to_signal(obs_pipewire_data *obs_pw,
+					   const char *path,
+					   GDBusSignalCallback callback)
+{
+	dbus_call_data *call;
+
+	call = g_new0(dbus_call_data, 1);
+	call->obs_pw = obs_pw;
+	call->request_path = g_strdup(path);
+	call->cancelled_id = g_signal_connect(obs_pw->cancellable, "cancelled",
+					      G_CALLBACK(on_cancelled_cb),
+					      call);
+	call->signal_id = g_dbus_connection_signal_subscribe(
+		obs_pw->connection, "org.freedesktop.portal.Desktop",
+		"org.freedesktop.portal.Request", "Response",
+		call->request_path, NULL, G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
+		callback, call, NULL);
+
+	return call;
+}
+
+static void dbus_call_data_free(dbus_call_data *call)
+{
+	if (!call)
+		return;
+
+	if (call->signal_id)
+		g_dbus_connection_signal_unsubscribe(call->obs_pw->connection,
+						     call->signal_id);
+
+	if (call->cancelled_id > 0)
+		g_signal_handler_disconnect(call->obs_pw->cancellable,
+					    call->cancelled_id);
+
+	g_clear_pointer(&call->request_path, g_free);
+	g_free(call);
+}
+
+static void teardown_pipewire(obs_pipewire_data *obs_pw)
+{
+	if (obs_pw->thread_loop) {
+		pw_thread_loop_wait(obs_pw->thread_loop);
+		pw_thread_loop_stop(obs_pw->thread_loop);
+	}
+
+	if (obs_pw->stream)
+		pw_stream_disconnect(obs_pw->stream);
+	g_clear_pointer(&obs_pw->stream, pw_stream_destroy);
+	g_clear_pointer(&obs_pw->context, pw_context_destroy);
+	g_clear_pointer(&obs_pw->thread_loop, pw_thread_loop_destroy);
+
+	if (obs_pw->pipewire_fd > 0) {
+		close(obs_pw->pipewire_fd);
+		obs_pw->pipewire_fd = 0;
+	}
+
+	obs_pw->negotiated = false;
+}
+
+static void destroy_session(obs_pipewire_data *obs_pw)
+{
+	if (obs_pw->session_handle) {
+		g_dbus_connection_call(
+			obs_pw->connection, "org.freedesktop.portal.Desktop",
+			obs_pw->session_handle,
+			"org.freedesktop.portal.Session", "Close", NULL, NULL,
+			G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL, NULL);
+
+		g_clear_pointer(&obs_pw->session_handle, g_free);
+	}
+
+	g_clear_pointer(&obs_pw->sender_name, g_free);
+	g_clear_pointer(&obs_pw->cursor.texture, gs_texture_destroy);
+	g_clear_pointer(&obs_pw->texture, gs_texture_destroy);
+	g_cancellable_cancel(obs_pw->cancellable);
+	g_clear_object(&obs_pw->cancellable);
+	g_clear_object(&obs_pw->connection);
+	g_clear_object(&obs_pw->proxy);
+}
+
+static inline bool has_effective_crop(obs_pipewire_data *obs_pw)
+{
+	return obs_pw->crop.valid &&
+	       (obs_pw->crop.x != 0 || obs_pw->crop.y != 0 ||
+		obs_pw->crop.width < obs_pw->format.info.raw.size.width ||
+		obs_pw->crop.height < obs_pw->format.info.raw.size.height);
+}
+
+static bool
+spa_pixel_format_to_obs_pixel_format(uint32_t spa_format,
+				     enum gs_color_format *out_format)
+{
+	switch (spa_format) {
+	case SPA_VIDEO_FORMAT_RGBA:
+	case SPA_VIDEO_FORMAT_RGBx:
+		*out_format = GS_RGBA;
```

Comment:
Hm. I wonder when this could happen in the Wayland world. These frames are provided by the compositor, and they shouldn't hand an RGBX buffer that's actually RGBA with noise in the alpha channel. If they do, that's probably a bug...

In any case, I've added code to swap R/B there, and if we receive RGBX, the texture is setup as BGRX with swapped R/B.

Reply:
How this happens in X11 I think: The window surface is RGBX but the game may do the final render copying from a RGBA texture expecting (correctly) for the extra alpha data they used to be ignored during compositing. But when we copy that texture back into our buffers and display it as RGBA we render the previously ignored alpha data.

I remember i bought rocketleague to test this so i can play with the old code and see if it happens, but I only have gnome on my test machine.

Reply:
i was playing with this and it appears that if you try to load the dmabuf with the wrong format it just wont render anything (and no errors output either). So RGBX may just not be supportable at the moment. At least on intel hardware and modern gnome now textures are BGRX as expected.

Reply:
plugged through drm formats and forced a render with the BGRX drm fd and obs as BGRA and couldnt replicate the issue so it may be a non-issue with drm. (but this means we RGBX will still fail unless we do something like plug through drm formats instead of trying to deduce from obs's texture formats since obs wont support RGBX since its not supported in directx)

Reply:
> I wonder when this could happen in the Wayland world. These frames are provided by the compositor, and they shouldn't hand an RGBX buffer that's actually RGBA with noise in the alpha channel.

I think there was a thread on wayland-devel recently with a question on how to do just that — have the alpha channel ignored by the compositor. Can't find it right now for some reason. But IIRC the conclusion was that you can't really do that currently, and it would require clearing out the alpha channel manually.

Reply:
well it still happens in modern gnome's xwayland, compiling the old v23 capture plugin you can still replicate the non-cleared alpha channel fun with rocketleague. So in that sense wayland rendering x11 is dropping the alpha data already. (Sorry I dont mean to getting too offtopic on this, so perhaps lets leave it there).